File: java/hamming/Hamming.java
/**
 * Hamming
 * https://exercism.org/tracks/java/exercises/hamming
 */
public class Hamming {

    private final int distance;

    public Hamming(String leftStrand, String rightStrand) {
        this.distance = calculateHammingDistance(leftStrand, rightStrand);
    }

    private int calculateHammingDistance(String leftStrand, String rightStrand) {
        if (leftStrand.isEmpty() && rightStrand.isEmpty()) {
            return 0;
        }

        if (leftStrand.isEmpty()) {
            throw new IllegalArgumentException("left strand must not be empty.");
        }

        if (rightStrand.isEmpty()) {
           throw new IllegalArgumentException("right strand must not be empty.");
        }

        if (leftStrand.length() != rightStrand.length()) {
            throw new IllegalArgumentException("leftStrand and rightStrand must be of equal length.");
        }

        int distance = 0;
        for (int i = 0; i < leftStrand.length(); i++) {
            if (leftStrand.charAt(i) != rightStrand.charAt(i)) {
                distance++;
            }
        }

        return distance;
    }

    public int getHammingDistance() {
        return distance;
    }
}

File: c/hamming/hamming.c
#include "hamming.h"

/**
 * Hamming
 * https://exercism.org/tracks/c/exercises/hamming
 */

int compute(const char *lhs, const char *rhs)
{
    int distance = 0;
    int i = 0;
    while(lhs[i] && rhs[i]) {
        if (lhs[i] != rhs[i]) {
            distance++;
        }

        i++;
    }

    if (lhs[i] != rhs[i]) {
        return -1;
    }

    return distance;
}
